### PR TITLE
Added Matt's proof for ceilLog2's CLZ implementation

### PR DESCRIPTION
**Pull request by djmorel**
# Features:
- Included a new directory, `labs/SAW/ceilLog2`
- Added Matt's source code and proof for the `CLZ` implementation of `ceilLog2`
- Added a Makefile in the `ceilLog2` directory to run the example
- Updated SAW.md to mention the CLZ verification as a bonus exercise

### CODE REVIEW

File: labs/SAW/ceilLog2/Makefile [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ARTIFACTS=artifacts
+
+# Note: -g flag compiles the code with debug symbols enabled.
+#       The flag lets us pass field names in addition to indices for structs.
+all: artifacts
+	clang -c -g -emit-llvm -o $(ARTIFACTS)/ceilLog2.bc $(SRC)/ceilLog2.c
+	python3 $(PROOF)/ceilLog2.py
+
+artifacts:
+	mkdir $(ARTIFACTS)
```
Comment (weaversa) [inline labs/SAW/ceilLog2/Makefile:17]: Better to test if the directory exists before creating it.

`[ -d artifacts ] || mkdir -p artifacts`

**Review by weaversa — COMMENTED** [review]
Looking great! Would you be willing to `make artifacts` in the CI -- .github/workflows/python-saw.yml ?

File: labs/SAW/ceilLog2/Makefile (hunk 1 above)
Comment (djmorel) [reply labs/SAW/ceilLog2/Makefile]: Fair point. I added the `-p` flag for Makefiles calling `mkdir`.

**Comment by djmorel** [comment]
> Looking great! Would you be willing to `make artifacts` in the CI -- .github/workflows/python-saw.yml ?

Sure thing! Based on the previously included steps in `python-saw.yml`, I wrote out the explicit steps rather than just calling `make` after navigating via `cd`. I take it that it is better to be explicit in the steps that way someone looking at just `python-saw.yml` can understand exactly what is being checked, right?

**Comment by djmorel** [comment]
Including this for future reference.

If you are pushing workflow changes in Gitpod and a "workflow scope" error appears like this:
```sh
gitpod /workspace/cryptol-course/labs/SAW (clz-proof) $ git push origin clz-proof 
Enumerating objects: 9, done.
Counting objects: 100% (9/9), done.
Delta compression using up to 16 threads
Compressing objects: 100% (4/4), done.
Writing objects: 100% (5/5), 475 bytes | 475.00 KiB/s, done.
Total 5 (delta 3), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (3/3), completed with 3 local objects.
To https://github.com/weaversa/cryptol-course.git
 ! [remote rejected] clz-proof -> clz-proof (refusing to allow an OAuth App to create or update workflow `.github/workflows/python-saw.yml` without `workflow` scope)
error: failed to push some refs to 'https://github.com/weaversa/cryptol-course.git'
```

Update your [Gitpod Integration permissions](https://gitpod.io/integrations) for GitHub. Click on the 3 dots for the GitHub permissions > select `Edit Permissions` > check the `workflow` permission > select `Update Permissions`. After doing so, GitHub will prompt you to confirm the permission changes and, after confirming them, you'll be able to push your commit. 
  
Gitpod_WorkflowsPermission

**Comment by weaversa** [comment]
> > Looking great! Would you be willing to `make artifacts` in the CI -- .github/workflows/python-saw.yml ?
> 
> Sure thing! Based on the previously included steps in `python-saw.yml`, I wrote out the explicit steps rather than just calling `make` after navigating via `cd`. I take it that it is better to be explicit in the steps that way someone looking at just `python-saw.yml` can understand exactly what is being checked, right?

I'd rather have the CI run a `make prove` command (see [here](https://github.com/weaversa/cryptol-course/blob/master/labs/Demos/SAW/xxHash/Makefile) for an example, though it doesn't use the python interface). I think this is better because if we decide to build differently (add some header file, change optimization flags, etc.) the proof will be linked to that build process. If you can come up with a good demonstration here, I'd like to change over all of the builds to work this way - we could still have two ci stages (if this is helpful, I'm not sure) - a `make proofartifacts` and a `make prove` or something...

File: labs/SAW/ceilLog2/Makefile (hunk 1 above)
Comment (weaversa) [reply labs/SAW/ceilLog2/Makefile]: Thanks!

**Comment by weaversa** [comment]
> > > Looking great! Would you be willing to `make artifacts` in the CI -- .github/workflows/python-saw.yml ?
> > 
> > 
> > Sure thing! Based on the previously included steps in `python-saw.yml`, I wrote out the explicit steps rather than just calling `make` after navigating via `cd`. I take it that it is better to be explicit in the steps that way someone looking at just `python-saw.yml` can understand exactly what is being checked, right?
> 
> I'd rather have the CI run a `make prove` command (see [here](https://github.com/weaversa/cryptol-course/blob/master/labs/Demos/SAW/xxHash/Makefile) for an example, though it doesn't use the python interface). I think this is better because if we decide to build differently (add some header file, change optimization flags, etc.) the proof will be linked to that build process. If you can come up with a good demonstration here, I'd like to change over all of the builds to work this way - we could still have two ci stages (if this is helpful, I'm not sure) - a `make proofartifacts` and a `make prove` or something...

I added the makefile syntax.

**Review by weaversa — APPROVED** [review]
Looks great! Thanks @djmorel !!!

**Comment by djmorel** [comment]
I like that Makefile setup! Very nice and clean. I'll add similar Makefiles for the other SAW labs.

**Pull request merged**